features/375-static-cdn/design.md

```diff
@@ -0,0 +1,77 @@
+# Static Asset Hosting
+
+## Background
+Developer would want to host static content (e.g. single-page-application files) on Skygear and serve them through
+high performance CDN.
+
+## Objectives
+- Developers should be able to upload static content to be hosted at specific path.
+- Hosted static content should be served & cached with CDN if possible.
+- SPA style routing should be supported.
+
+## Design
+
+### Static Deployment
+
+A new type of deployment, `static`, would be supported. This deployment type
+supports configurating a deployment context (`context`) and fallback rewrite
+path (`fallback`).
+
+When `static` deployment type is used:
+- The deployment context would be persisted to object storage.
+- A request with matching path would be routing using static deployment routing
+  algorithm (described below).
+- If a file is matched, the gateway would proxy the request to the backing
+  object storage file.
+- If no file is matched, a generic 404 page would be returned.
+- The response would have appropiate cache control header, allowing maximum
```

Comment:
A typical deployment of single page application looks like

```
index.html
assets/js/main.688787d8ff144c502c7f5cffaafe2cc588d86079f9de88304c26b0cb99ce91c6.js
assets/css/main.cc00113f2c36b6d0e501dd832771ce5db7b41c1d77edb22b5fde8baf1c5e449b.css
```

where we want to cache the files in `assets/` for a very long period time while cache `index.html` for a very short period time. How can we support this?

Reply:
For now, we have a fixed cache duration of 24 hours, so changes to `index.html` would takes 24 hours to roll out completely. I'm not sure if how flexible we should allow configuring cache duration, since misconfiguration can cause prolonged stale cached content/performance issues.

To support your use case, we can have cache duration configuration on static deployment, so developer can have two static deployment:
- `path=/, context=./build, fallback=/index.html, duration=1h`
- `path=/assets, context=./build/assets, duration=30d`

Do you think it is flexible enough?

Reply:
This design allows a path-prefix based caching configuration and should be good enough. We may follow nginx to name `duration` `expires`. I guess there is no cache control headers if the developer does not specify it, right?

Reply:
I think we'll have a sensible default of 1 hour cache duration. Developer can disable caching by setting it explicitly to 0, but the maximum cache duration is 7 days.

Reply:
Added `expires` configuration.

Reply:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control

The related directives are `public`, `no-cache`, `must-revalidate` and `immutable`. For path that contains hash, we should specify `public, immutable, max-age=<max-age>`.
For index.html, we should specify `public, no-cache, max-age=<max-age>` or `public, must-revalidate, max-age=<max-age>`. Should we support this as well?

Reply:
- `immutable`: It seems quite dangerous. If I understand correctly, it would poison the cache forever if used incorrectly, therefore I don't suggest supporting this.
- `no-cache`/`must-validate`: I think `no-cache` would be used automatically if duration is zero, otherwise `must-validate` would be used.

This should cover most use cases without major problems. I'll update to include the concrete injected cache control headers. Do you have any other suggestions?